### ISSUE TITLE
Some questions about implementation of OHEM cross entropy loss

### ISSUE THREAD

**Issue by mvish7** [DeyvidKochanov-TomTom/kprnet#13]
Hello,

I would like to get some clarification about the implementation of OHEM loss. I'm having difficulties in reasoning the implementation done in the following lines.

```
        tmp_target = target.clone()
        tmp_target[tmp_target == self.ignore_index] = 0
        pred = pred.gather(1, tmp_target.unsqueeze(1))
        pred, ind = pred.contiguous().view(-1,)[mask].contiguous().sort()
        min_value = pred[min(self.min_kept, pred.numel() - 1)]
        threshold = max(min_value, self.thresh)

        pixel_losses = pixel_losses[mask][ind]
        pixel_losses = pixel_losses[pred < threshold]
```
Especially, I would like to understand what is the intention behind this line?? 
-->>` pred = pred.gather(1, tmp_target.unsqueeze(1))`

I would also like to ask you the following questions:

1. How did you derive the `min_kept` and `threshold` parameters?? 
2. Did you try using the `weight` option in nn.CrossEntropyLoss??

Thanks in advance

**Comment by DeyvidKochanov-TomTom** [comment]
Hi,
the line gathers the probabilities of the correct label for each pixel, I think. You could run the line and find out :P
As for the parameters, I haven't tuned them so they are the same as in HRNet with just the proportion of the kept pixels roughly the same. 
I think the per-class weights are overall very hacky but you could try it! I think it leads to a small improvement with the OHEM loss.

**Comment by mvish7** [comment]
hey,
Thanks for the reply. Below I have given my understanding of OHEM loss implementation, by doing so I have tried to pin point my question. Please feel free to correct me. 

`mask = target.contiguous().view(-1) != self.ignore_index`
The above line creates a Binary tensor that has a False at each place for the value=ignore_index

`tmp_target[tmp_target == self.ignore_index] = 0`
tmp_target is a copy of GT and This line replaces each ignore_index value by a 0.

`pred = pred.gather(1, tmp_target.unsqueeze(1))`
This line gathers values from the 1st dimension of pred tensor and the indices from which to pick the values are given by tmp_target.

##Now here I get confused. What I'm thinking is as follows:
>>>>>>>
we just replaced ignore_index by 0 in tmp_target and before that tmp_target already had 0s for classes like 'car'. As seen [Here](https://github.com/DeyvidKochanov-TomTom/kprnet/blob/587a616b955b6be55b223e031e9ed39a8588d149/datasets/semantic_kitti.py#L202)

So `pred = pred.gather(1, tmp_target.unsqueeze(1))` this line is picking up the value from 0th index for class 'car' and also for class represeted by ignore_index. 
>>>>>>>>>>

Am I right here or am I getting something very wrong???

**Comment by DeyvidKochanov-TomTom** [comment]
Yes there is one label per pixel so for both class 0 and the ignore class the gather will collect the probability at index 0,  but later the mask  tensor is used on pred so the values where we switched the label from ignore to 0  are ignored.  Its just so the gather call doesn't overflow when ignore index is 255 and your tensor size in dim 1 is 19 or whatever:

```
      pred = pred.gather(1, tmp_target.unsqueeze(1))
      pred, ind = pred.contiguous().view(-1,)[mask].contiguous().sort()
```

**Comment by mvish7** [comment]
Hi,
Thanks for the clarification. My bad that I missed the masking of pred tensor.
